metadata-ingestion/src/datahub/ingestion/source/sql/db2.py

```diff
@@ -0,0 +1,298 @@
+)
+class Db2Source(SQLAlchemySource):
+    def __init__(self, config: Db2Config, ctx: PipelineContext):
+        super().__init__(config, ctx, "db2")
+        # register custom SQLGlot dialect
+        if not sqlglot.Dialect.get("db2"):
+            sqlglot.Dialect.classes["db2"] = CustomDb2SqlGlotDialect
+
+    def get_inspectors(self) -> Iterable[Inspector]:
+        for inspector in super().get_inspectors():
+            # use our custom SQLAlchemy dialect
+            patch_dialect(inspector.dialect)
+            yield inspector
+
+    def get_db_name(self, inspector: Inspector) -> str:
+        # database names are case-insensitive, so normalize them to uppercase
+        # to match everything else.
+        return super().get_db_name(inspector).upper()
+
+    def get_identifier(
+        self, *, schema: str, entity: str, inspector: Inspector, **kwargs: Any
+    ) -> str:
+        # database.schema.object
+        return f"{self.get_db_name(inspector)}.{schema}.{entity}"
+
+    def get_view_default_db_schema(
+        self, _dataset_name: str, inspector: Inspector, schema: str, view: str
+    ) -> Tuple[Optional[str], Optional[str]]:
+        # Db2 views on LUW and z/OS look up unqualified names in the schema from the session
+        # when the view was created. Not the schema that the view itself lives in!
+        schema_name = _db2_get_view_qualifier_quoted(inspector, schema, view)
+        db_name = self.get_db_name(inspector)
+        return db_name, schema_name
+
+    def get_procedures_for_schema(
+        self, inspector: Inspector, schema: str, _db_name: str
+    ) -> Iterable[BaseProcedure]:
+        for row in _db2_get_procedures(inspector, schema):
+            if row["QUALIFIER"]:
+                # similar to views, stored procedures can look up unqualified names in the
+                # schema from the session when they were created, not the schema of the
+                # proc itself. also quote the schema name so case-sensitive names make
+                # it through sqlglot without being normalized.
+                default_schema = _quote_identifier(row["QUALIFIER"].rstrip())
+            else:
+                default_schema = None
+
+            yield BaseProcedure(
+                name=row["ROUTINENAME"],
+                # language can have trailing spaces
+                language=row["LANGUAGE"].rstrip() if row["LANGUAGE"] else "",
+                procedure_definition=row["TEXT"],
+                comment=row["REMARKS"],
+                created=row["CREATE_TIME"],
+                last_altered=row["ALTER_TIME"],
+                default_schema=default_schema,
+                argument_signature=None,
+                return_type=None,
+                extra_properties={},
+            )
+
+
+def _db2_get_view_qualifier_quoted(
+    inspector: Inspector, schema: str, view: str
+) -> Optional[str]:
+    if inspector.has_table("VIEWS", schema="SYSCAT"):
```

Comment:
we may define an enum for: Db2 LUW, Db2 z/OS and Db2 i/AS400

and a method: def _identify_db2_type(inspector): enum

because I would expect that to become a recurrent pattern

Reply:
I don't agree on this one. Adding another method adds an additional layer of indirection and something to break, and we'd have to decide upon a way to do the discrimination (check the underlying SQLAlchemy inspector reflector type?). The current way is straightforward and doesn't have any business logic issues: if table X exists, use it.